File: week16/siu/257.go
package siu

import "fmt"

//给定一个二叉树，返回所有从根节点到叶子节点的路径。
//
// 说明: 叶子节点是指没有子节点的节点。
//
// 示例:
//
// 输入:
//
//   1
// /   \
//2     3
// \
//  5
//
//输出: ["1->2->5", "1->3"]
//
//解释: 所有根节点到叶子节点的路径为: 1->2->5, 1->3
// Related Topics 树 深度优先搜索
// 👍 416 👎 0

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func binaryTreePaths(root *TreeNode) []string {
	res := []string{}
	dfs(root, "", &res)
	return res
}

func dfs(node *TreeNode, pathStr string, res *[]string) {
	if node == nil {
		return
	}

	if node.Left == nil && node.Right == nil {
		pathStr += fmt.Sprintf("%d", node.Val)
		*res = append(*res, pathStr)
		return
	}

	pathStr += fmt.Sprintf("%d->", node.Val)
	dfs(node.Left, pathStr, res)
	dfs(node.Right, pathStr, res)
}

//leetcode submit region end(Prohibit modification and deletion)

File: week16/Aiden/257.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        def binaryTreePaths(root: TreeNode) -> List[str]:
            if not root:
                return []  # 如果当前是空则另一边是一棵二叉树，所以他的祖先不可能是叶子节点。
            if not root.left and not root.right:
                return [str(root.val)]  # 递归到这里有一条路径。
            return [str(root.val)+"->"+i for i in binaryTreePaths(root.left)+binaryTreePaths(root.right)]
        return binaryTreePaths(root)